File: BMM.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <unordered_set>
#include <algorithm>

// 读取词典文件并存储为一个unordered_set
std::unordered_set<std::string> loadDictionary(const std::string& dictionaryFile) {
    std::unordered_set<std::string> wordDict;
    std::ifstream file(dictionaryFile);
    if (file.is_open()) {
        std::string word;
        while (std::getline(file, word)) {
            wordDict.insert(word);
        }
        file.close();
    } else {
        std::cerr << "Error: Unable to open dictionary file." << std::endl;
    }
    return wordDict;
}

// 使用双向最大匹配法进行分词
std::vector<std::string> bmmSegment(const std::string& text, const std::unordered_set<std::string>& wordDict) {
    std::vector<std::string> result;
    int maxWordLength = 0;
    for (const std::string& word : wordDict) {
        if (word.length() > maxWordLength) {
            maxWordLength = word.length();
        }
    }

    int textLength = text.length();
    int i = 0;

    while (i < textLength) {
        int length = std::min(maxWordLength, textLength - i);
        while (length > 0) {
            std::string sub = text.substr(i, length);
            if (wordDict.find(sub) != wordDict.end()) {
                result.push_back(sub);
                i += length;
                break;
            }
            length--;
        }

        if (length == 0) {
            length = 1;
            result.push_back(text.substr(i, length));
            i++;
        }
    }

    return result;
}

int main() {
    // 替换成你的词典文件路径
    std::string dictionaryFile = "dictionary.txt";
    std::string text = "我喜欢编程和学习";

    std::unordered_set<std::string> wordDict = loadDictionary(dictionaryFile);
    std::vector<std::string> result = bmmSegment(text, wordDict);

    for (const std::string& word : result) {
        std::cout << word << " ";
    }

    return 0;
}

File: BMM.py
import yaml

# 读取词典文件
def load_dictionary(dictionary_file):
    with open(dictionary_file, 'r', encoding='utf-8') as file:
        word_dict = yaml.load(file, Loader=yaml.FullLoader)
    return set(word_dict)

# 使用双向最大匹配法进行分词
def bmm_segment(text, word_dict):
    result = []
    max_word_length = max(len(word) for word in word_dict)

    text_length = len(text)
    i = 0

    while i < text_length:
        length = min(max_word_length, text_length - i)
        while length > 0:
            sub = text[i:i + length]
            if sub in word_dict:
                result.append(sub)
                i += length
                break
            length -= 1

        if length == 0:
            length = 1
            result.append(text[i:i + length])
            i += length

    return result

# 主程序
if __name__ == '__main__':
    # 替换成你的词典文件路径
    dictionary_file = "corpuswordlist.dict.yaml"

    # 替换成你的待分词文本文件路径
    text_file = "text.txt"

    word_dict = load_dictionary(dictionary_file)

    with open(text_file, 'r', encoding='utf-8') as file:
        text = file.read()

    result = bmm_segment(text, word_dict)
    print("分词结果:", " ".join(result))
